### PR TITLE
Create AutoRenewable STS/SAML credentialsProvider

### PR DESCRIPTION
**Pull request by mscharp**
Added new Sts credential provider class that auto renews a SAML assertion
Created interface that can be implemented to provide functionality to CredentialProvider

## Motivation and Context
When using the StsAssumeRoleWithSamlCredentialProvider, a non expired saml assertion is needed before the AWS credentials can be refreshed.  Currently, the SDK doesn't provide a way to handle renewing the saml assertion before calling the STS service to obtain new AWS credentials.  This pull requests implements a generic solution for this.

It's possible that the new class could be merged into the current StsAssumeRoleWithSamlCredentialProvider, but I wasn't sure if that would be the best way to handle it.

## Testing
Added unit tests for the new class. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ X] New feature (non-breaking change which adds functionality)

## Checklist

- [ X ] I have read the **CONTRIBUTING** document
- [ X] Local run of `mvn install` succeeds
- [ X ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ X ] I have read the **README** document
- [ X ] I have added tests to cover my changes
- [ X ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

- [ X ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by millems** [comment]
+1 to merging this with the existing credentials provider instead of creating a new one. Instead of reactively updating the assertion, can we proactively update it when we create the request?

We've added something similar to the assume-role credentials provider that allows specifying a request OR a request provider (for use with MFA): https://github.com/aws/aws-sdk-java-v2/blob/master/services/sts/src/main/java/software/amazon/awssdk/services/sts/auth/StsAssumeRoleCredentialsProvider.java#L99-L109

If a similar pattern would work here, that seems ideal.

**Comment by mscharp** [comment]
I can merge the changes into the existing provider easily enough.  I'll have an update shortly.

Could you elaborate more on how the request provider would work?

I initially created a solution that would keep track of the expiry time and proactively reauthenticate with the saml, but I didn't like the solution.  It required creating a timer and trying to time the saml reauthentication with the aws reauthentication, and over time got out of sync.  This may not have been the best way to do it, but it was what I had come up with.  

As such, I went with this approach as it shouldn't impact performance or usability, and has the added benefit of being quite simple.

**Comment by mscharp** [comment]
Ok.  I took a deeper look and now understand how the Supplier works. (I haven't used that pattern before).  I think that will work great.  It's similar to what I already implemented, but executes a little sooner (before the ETE is thrown) and doesn't require a timer or custom interface.  Much cleaner!

I should have it updated soon.

**Comment by millems** [comment]
Awesome, it'd be great to have that functionality added to the default SAML provider.

**Comment by mscharp** [comment]
Code has been updated and rebased against master.  Just waiting for review now.

File: .changes/next-release/feature-AWSSecureTokenService-6de82b5.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "category": "AWS SecureTokenService",
```
Comment (dagnir) [inline .changes/next-release/feature-AWSSecureTokenService-6de82b5.json]: minor: This should be 'AWS Security Token Service'

**Comment by mscharp** [comment]
I've rebased my feature branch with master and squashed the commits.

**Comment by mscharp** [comment]
Bumping to hopefully get it merged in soon...

File: .changes/next-release/feature-AWSSecureTokenService-6de82b5.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+    "category": "AWS Secure Token Service",
```
Comment (dagnir) [inline .changes/next-release/feature-AWSSecureTokenService-6de82b5.json]: nit: this should be "Security"

**Comment by mscharp** [comment]
@dagnir 

Rebased and fixed.

**Comment by dagnir** [comment]
@mscharp Thanks for your contribution!

**Pull request merged**